src/Interpreters/Cache/FileCache.cpp

```diff
@@ -887,149 +863,217 @@ void FileCache::loadMetadata()
             "Please, check log for error messages");
     }
 
-    size_t total_size = 0;
-    for (auto key_prefix_it = fs::directory_iterator{metadata.getBaseDirectory()}; key_prefix_it != fs::directory_iterator();
-         key_prefix_it++)
-    {
-        const fs::path key_prefix_directory = key_prefix_it->path();
+    loadMetadataImpl();
+
+    /// Shuffle file_segment_metadatas to have random order in LRUQueue
+    /// as at startup all file_segment_metadatas have the same priority.
+    main_priority->shuffle(lockCache());
+}
 
-        if (!key_prefix_it->is_directory())
+void FileCache::loadMetadataImpl()
+{
+    auto get_keys_dir_to_process = [
+        &, key_prefix_it = fs::directory_iterator{metadata.getBaseDirectory()}, get_key_mutex = std::mutex()]
+        () mutable -> std::optional<fs::path>
+    {
+        std::lock_guard lk(get_key_mutex);
+        while (true)
         {
-            if (key_prefix_directory.filename() != "status")
+            if (key_prefix_it == fs::directory_iterator())
+                return std::nullopt;
+
+            auto path = key_prefix_it->path();
+            if (key_prefix_it->is_directory())
             {
-                LOG_WARNING(
-                    log, "Unexpected file {} (not a directory), will skip it",
-                    key_prefix_directory.string());
+                key_prefix_it++;
+                return path;
+            }
+
+            if (key_prefix_it->path().filename() != "status")
+            {
+                LOG_WARNING(log, "Unexpected file {} (not a directory), will skip it", path.string());
             }
+            key_prefix_it++;
+        }
+    };
+
+    std::vector<ThreadFromGlobalPool> loading_threads;
+    std::exception_ptr first_exception;
+    std::mutex set_exception_mutex;
+    std::atomic<bool> stop_loading = false;
+
+    LOG_INFO(log, "Loading filesystem cache with {} threads", metadata_download_threads);
+
+    for (size_t i = 0; i < metadata_download_threads; ++i)
+    {
+         loading_threads.emplace_back([&]
+         {
+             while (!stop_loading)
+             {
+                 try
+                 {
+                     auto path = get_keys_dir_to_process();
+                     if (!path.has_value())
+                         return;
+
+                     loadMetadataForKeys(path.value());
+                 }
+                 catch (...)
+                 {
+                     {
+                        std::lock_guard exception_lock(set_exception_mutex);
+                        if (!first_exception)
+                            first_exception = std::current_exception();
+                     }
+                     stop_loading = true;
+                     return;
+                 }
+             }
+         });
+    }
+
+    for (auto & thread : loading_threads)
```

Comment:
in case of exception in the loop we will not wait for threads

Reply:
Same as in  `ThreadPool.cpp`: https://github.com/ClickHouse/ClickHouse/blob/c3c41253281bd3fb7f9209ac9bb4482db6597d6f/src/Common/ThreadPool.cpp#L295-L296
also by the code of `ThreadFromGlobalPool` it does not look like `join()` can throw an exception